File: backend/installed-extensions/manhwalatino.js
import { BaseExtension } from '../base/BaseExtension.js';
import axios from 'axios';
import * as cheerio from 'cheerio';

const GENRE_MAP = {
  "+18": "351",
  "+19": "309",
  "academia": "183",
  "acción": "64",
  "adulto": "92",
  "aliens": "203",
  "amigos con derechos": "134",
  "amigos de la infancia": "142",
  "anti-heroe": "349",
  "apocaliptico": "156",
  "apocalíptico": "156",
  "artes marciales": "81",
  "aventura": "65",
  "bestias": "343",
  "ciencia ficción": "85",
  "comedia": "67",
  "crimen": "251",
  "cultivo": "166",
  "cyberpunk": "86",
  "deporte": "219",
  "deportes": "219",
  "doujinshi": "322",
  "drama": "75",
  "ecchi": "68",
  "ejercito": "176",
  "fantasía": "69",
  "girls love": "162",
  "gore": "177",
  "guerra": "144",
  "harem": "88",
  "historias cortas": "190",
  "horror": "70",
  "isekai": "97",
  "josei": "169",
  "madrastra": "197",
  "madre e hija": "172",
  "maduro": "352",
  "magia": "98",
  "manwha": "310",
  "manhwa": "310",
  "milfs": "103",
  "misterio": "83",
  "mujer casada": "104",
  "musica": "212",
  "música": "212",
  "pimer amor": "135",
  "primer amor": "189",
  "psicológico": "82",
  "recuentos de la vida": "117",
  "reencarnación": "99",
  "relacion secreta": "119",
  "relación secreta": "119",
  "romance": "89",
  "seinen": "138",
  "shoujo": "93",
  "shounen": "78",
  "sistema": "100",
  "sobrenatural": "101",
  "superpoderes": "257",
  "tragedia": "80",
  "universidad": "106",
  "vampiros": "157",
  "venganza": "130",
  "vida escolar": "299",
  "yuri": "161"
};

export class ManhwaLatinoExtension extends BaseExtension {
  constructor(config = {}) {
    super({
      id: config.id || 'manhwalatino',
      name: config.name || 'ManhwaLatino',
      version: config.version || '2.5.0',
      baseUrl: config.baseUrl || 'https://manhwalatino.lat',
      icon: config.icon || 'https://manhwalatino.lat/favicon.ico',
      lang: config.lang || 'es'
    });
    this.headers = {
      'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
      'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8',
      'Accept-Language': 'es-ES,es;q=0.9,en;q=0.8'
    };
  }

  _buildSearchUrl(page = 1, filters = {}) {
    const params = new URLSearchParams();
    if (page > 1) {
      params.append('page', String(page));
    }

    if (filters.query && filters.query.trim()) {
      params.append('title', filters.query.trim());
    }

    if (filters.status && filters.status !== 'Ver todo') {
      const s = filters.status.toLowerCase();
      if (s.includes('activo') || s.includes('emisión')) params.append('status', 'ongoing');
      else if (s.includes('finalizado') || s.includes('completado')) params.append('status', 'completed');
    }

    if (filters.type && filters.type !== 'Ver todo') {
      const t = filters.type.toLowerCase();
      if (t.includes('manhwa')) params.append('type', 'manhwa');
      else if (t.includes('manhua')) params.append('type', 'manhua');
      else if (t.includes('manga')) params.append('type', 'manga');
    }

    const sort = filters.sort || 'popular';
    if (sort === 'reciente' || sort === 'recent' || sort === 'latest') {
      params.append('order', 'update');
    } else if (sort === 'rating' || sort === 'valoracion') {
      params.append('order', 'rating');
    } else if (sort === 'az' || sort === 'title' || sort === 'alfabetico') {
      params.append('order', 'title');
    } else {
      params.append('order', 'popular');
    }

    if (filters.genres && filters.genres.length > 0) {
      for (const g of filters.genres) {
        const gId = GENRE_MAP[g.toLowerCase().trim()];
        if (gId) {
          params.append('genre[]', gId);
        }
      }
    }

    const qs = params.toString();
    return `${this.baseUrl}/manga/${qs ? '?' + qs : ''}`;
  }

  async _fetchRawPage(page = 1, filters = {}) {
    const url = this._buildSearchUrl(page, filters);
    try {
      const res = await axios.get(url, { headers: this.headers, timeout: 12000 });
      const html = res.data;
      const $ = cheerio.load(html);
      const mangas = [];

      $('.bsx').each((_, el) => {
        const link = $(el).find('a').first();
        const href = link.attr('href');
        const title = $(el).find('.tt, [itemprop="headline"]').text().trim() || link.attr('title') || '';
        const img = $(el).find('img').first();
        const cover = img.attr('data-src') || img.attr('data-lazy-src') || img.attr('src') || '';
        const latestChapter = $(el).find('.epxs').text().trim() || 'Disponible';
        const score = $(el).find('.numscore').text().trim();
        const type = $(el).find('.type').text().trim() || 'Manhwa';

        let status = 'Activo';
        const statusText = $(el).find('.status, .status_tag').text().trim().toLowerCase();
        if (statusText.includes('complet') || statusText.includes('final')) status = 'Finalizado';

        if (href && title && !mangas.some(m => m.url === href)) {
          mangas.push({
            id: href.replace(/\/+$/, '').split('/').pop(),
            title,
            url: href,
            cover,
            latestChapter,
            score: score ? `⭐ ${score}` : '',
            type,
            status
          });
        }
      });

      return mangas;
    } catch (err) {
      console.warn(`[ManhwaLatino] Error fetching raw page ${page}:`, err.message);
      return [];
    }
  }

  async getCatalog(page = 1, filters = {}) {
    try {
      const p = Math.max(1, parseInt(page, 10) || 1);
      const targetCount = 35; // 35 mangas por página normalizado
      const sourceCountPerPage = 30; // ManhwaLatino entrega 30 mangas por página fuente

      const startIdx = (p - 1) * targetCount;
      const endIdx = p * targetCount;
      const startPage = Math.floor(startIdx / sourceCountPerPage) + 1;
      const endPage = Math.floor((endIdx - 1) / sourceCountPerPage) + 1;

      const pagePromises = [];
      for (let sp = startPage; sp <= endPage; sp++) {
        pagePromises.push(this._fetchRawPage(sp, filters));
      }
      const results = await Promise.all(pagePromises);
      const combined = results.flat();
      const offset = startIdx - (startPage - 1) * sourceCountPerPage;
      const mangas = combined.slice(offset, offset + targetCount);

      const hasNext = combined.length > offset + mangas.length || mangas.length === targetCount;

      // ManhwaLatino tiene 209 títulos (7 páginas de 30 = 6 páginas de 35)
      const totalCount = 209;
      const totalPages = Math.max(1, Math.ceil(totalCount / targetCount));

      return {
        success: true,
        extension: this.name,
        extensionId: this.id,
        page: p,
        mangas,
        hasNextPage: p < totalPages && hasNext,
        totalPages,
        totalCount
      };
    } catch (err) {
      console.error('[ManhwaLatino Error] getCatalog:', err.message);
      return {
        success: false,
        extension: this.name,
        extensionId: this.id,
        page,
        mangas: [],
        hasNextPage: false,
        totalPages: 1,
        totalCount: 0
      };
    }
  }

  async search(query, page = 1, filters = {}) {
    return this.getCatalog(page, { ...filters, query });
  }

  async getMangaDetails(mangaUrl) {
    try {
      const res = await axios.get(mangaUrl, { headers: this.headers, timeout: 12000 });
      const html = res.data;
      const $ = cheerio.load(html);

      const title = $('h1.entry-title, .entry-title').first().text().trim();
      const cover = $('.thumb img, .entry-content img').first().attr('data-src') || $('.thumb img').attr('src') || '';
      const synopsis = $('.entry-content p, .sinopsis, [itemprop="description"]').first().text().trim() || 'Sin sinopsis disponible.';
      const author = $('.tsinfo .imptdt:contains("Autor") i, .tsinfo .imptdt:contains("Author") i').text().trim() || '';
      
      let status = 'Activo';
      const statusRaw = $('.tsinfo .imptdt:contains("Status") i, .tsinfo .imptdt:contains("Estado") i').text().trim().toLowerCase();
      if (statusRaw.includes('complet') || statusRaw.includes('final')) status = 'Finalizado';

      const typeRaw = $('.tsinfo .imptdt:contains("Type") i, .tsinfo .imptdt:contains("Tipo") i').text().trim();
      const type = typeRaw || 'Manhwa';

      const genres = $('.mgen a, .genres-content a').map((_, a) => $(a).text().trim()).get();

      const chapters = [];
      $('#chapterlist li, .eplister li').each((_, li) => {
        const a = $(li).find('a').first();
        const href = a.attr('href');
        const chNum = $(li).find('.chapternum').text().trim() || a.text().trim();
        const date = $(li).find('.chapterdate').text().trim();

        if (href && !chapters.some(c => c.url === href)) {
          chapters.push({
            name: chNum,
            chapterNumber: chNum.replace(/[^\d.]/g, '') || '1',
            url: href,
            date
          });
        }
      });

      return {
        id: mangaUrl.replace(/\/+$/, '').split('/').pop(),
        title,
        url: mangaUrl,
        cover,
        synopsis,
        author,
        status,
        type,
        genres,
        chapters
      };
    } catch (err) {
      console.error('[ManhwaLatino Error] getMangaDetails:', err.message);
      return {
        id: 'fallback',
        title: 'ManhwaLatino Manga',
        url: mangaUrl,
        cover: '',
        synopsis: 'Manga sincronizado desde ManhwaLatino.',
        status: 'Activo',
        genres: ['Manhwa'],
        chapters: []
      };
    }
  }

  async extractChapter(chapterUrl) {
    try {
      const res = await axios.get(chapterUrl, { headers: this.headers, timeout: 12000 });
      const html = res.data;
      const $ = cheerio.load(html);

      let imageUrls = [];

      // 1. Extraer desde payload oficial JavaScript ts_reader
      const tsMatch = html.match(/ts_reader\.run\((.*?)\);/s);
      if (tsMatch) {
        try {
          const json = JSON.parse(tsMatch[1]);
          if (json.sources?.[0]?.images) {
            imageUrls = json.sources[0].images;
          }
        } catch (e) {}
      }

      // 2. Fallback HTML selector
      if (imageUrls.length === 0) {
        $('#readerarea img, .reading-content img, .page-break img').each((_, img) => {
          const src = $(img).attr('data-src') || $(img).attr('data-lazy-src') || $(img).attr('src') || '';
          if (src && !src.includes('logo') && !src.includes('banner')) {
            imageUrls.push(src.trim());
          }
        });
      }

      const pages = imageUrls.map((url, idx) => ({
        index: idx + 1,
        url
      }));

      return {
        success: true,
        extension: this.name,
        extensionId: this.id,
        mangaTitle: 'ManhwaLatino',
        chapterTitle: 'Capítulo',
        currentUrl: chapterUrl,
        totalPages: pages.length,
        pages
      };
    } catch (err) {
      console.error('[ManhwaLatino Error] extractChapter:', err.message);
      return {
        success: false,
        extension: this.name,
        extensionId: this.id,
        mangaTitle: 'ManhwaLatino',
        chapterTitle: 'Capítulo',
        currentUrl: chapterUrl,
        totalPages: 0,
        pages: []
      };
    }
  }
}

export default ManhwaLatinoExtension;

File: ext-modules/es-manhwalatino/src/main/kotlin/org/yomori/extension/es/manhwalatino/ManhwaLatino.kt
package org.yomori.extension.es.manhwalatino

import eu.kanade.tachiyomi.network.GET
import eu.kanade.tachiyomi.source.model.Filter
import eu.kanade.tachiyomi.source.model.FilterList
import eu.kanade.tachiyomi.source.model.MangasPage
import eu.kanade.tachiyomi.source.model.Page
import eu.kanade.tachiyomi.source.model.SChapter
import eu.kanade.tachiyomi.source.model.SManga
import eu.kanade.tachiyomi.source.online.HttpSource
import eu.kanade.tachiyomi.util.asJsoup
import okhttp3.Headers
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.Request
import okhttp3.Response
import org.jsoup.nodes.Document
import java.text.SimpleDateFormat
import java.util.Locale

class ManhwaLatino : HttpSource() {

    override val name = "Manhwa Latino"
    override val baseUrl = "https://manhwalatino.lat"
    override val lang = "es"
    override val supportsLatest = true

    override fun headersBuilder(): Headers.Builder = Headers.Builder()
        .add("Referer", "$baseUrl/")
        .add("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8")
        .add("Accept-Language", "es-ES,es;q=0.9")

    private val dateFormat by lazy {
        SimpleDateFormat("MMMM d, yyyy", Locale("es"))
    }

    override fun popularMangaRequest(page: Int): Request {
        return GET("$baseUrl/manga/?page=$page&order=popular", headers)
    }

    override fun popularMangaParse(response: Response): MangasPage {
        val document = response.asJsoup()
        val elements = document.select("div.bsx, div.animposx, .manga-card, article.bs, .element, .page-item-detail")
        val mangas = mutableListOf<SManga>()
        val seen = mutableSetOf<String>()

        for (el in elements) {
            val link = el.selectFirst("a[href*='/manga/']") ?: continue
            val href = link.attr("abs:href").ifEmpty { link.attr("href") }
            if (href.isEmpty() || href == "$baseUrl/manga/" || href == "$baseUrl/manga" || href.contains("/page/") || href.endsWith("/manga/list-mode") || !seen.add(href)) continue

            val title = link.attr("title").ifEmpty {
                el.selectFirst(".tt, .title, .entry-title, h4, h3")?.text()?.trim() ?: link.text().trim()
            }
            if (title.isEmpty() || title.length <= 1) continue

            val img = el.selectFirst("img")
            var cover = img?.attr("abs:data-src")?.ifEmpty { img.attr("abs:data-lazy-src")?.ifEmpty { img.attr("abs:src")?.ifEmpty { img.attr("src") } } } ?: ""
            if (cover.startsWith("data:image")) {
                cover = img?.attr("abs:data-src")?.ifEmpty { img.attr("abs:data-lazy-src") } ?: ""
            }
            if (cover.startsWith("//")) {
                cover = "https:$cover"
            } else if (cover.startsWith("/")) {
                cover = "$baseUrl$cover"
            }

            val capMatch = Regex("""(?:Cap[íi]tulo|Cap\.?|Ep\.?)\s*(\d+(?:\.\d+)?)""", RegexOption.IGNORE_CASE).find(el.text())
            val capDesc = if (capMatch != null) "Capítulo ${capMatch.groupValues[1]}" else ""

            val manga = SManga.create().apply {
                setUrlWithoutDomain(href)
                this.title = title
                this.thumbnail_url = cover
                this.description = capDesc
                this.initialized = true
            }
            mangas.add(manga)
        }

        val hasNext = document.selectFirst("a[rel=next], .next, .pagination-next, a.r") != null || mangas.size >= 15
        return MangasPage(mangas, hasNext)
    }

    override fun latestUpdatesRequest(page: Int): Request {
        return GET("$baseUrl/manga/?page=$page&order=update", headers)
    }

    override fun latestUpdatesParse(response: Response): MangasPage = popularMangaParse(response)

    override fun searchMangaRequest(page: Int, query: String, filters: FilterList): Request {
        val url = "$baseUrl/manga/".toHttpUrl().newBuilder()
            .addQueryParameter("page", page.toString())

        if (query.isNotBlank()) {
            url.addQueryParameter("title", query.trim())
        }

        var orderVal = "popular"
        for (filter in filters) {
            when (filter) {
                is OrderByFilter -> orderVal = filter.toUriPart()
                is StatusFilter -> {
                    val status = filter.toUriPart()
                    if (status.isNotEmpty()) url.addQueryParameter("status", status)
                }
                is TypeFilter -> {
                    val type = filter.toUriPart()
                    if (type.isNotEmpty()) url.addQueryParameter("type", type)
                }
                is GenreFilter -> {
                    val genre = filter.toUriPart()
                    if (genre.isNotEmpty()) url.addQueryParameter("genre[]", genre)
                }
                else -> {}
            }
        }
        url.addQueryParameter("order", orderVal)

        return GET(url.build().toString(), headers)
    }

    override fun searchMangaParse(response: Response): MangasPage = popularMangaParse(response)

    override fun mangaDetailsParse(response: Response): SManga {
        val document = response.asJsoup()
        return SManga.create().apply {
            title = document.select("h1.entry-title, h1, .post-title h1").firstOrNull()?.text()?.trim()
                ?.replace(Regex("""&#8211;"""), "-") ?: "Manhwa Latino"
            
            val img = document.select("div.thumb img, .summary_image img, img.wp-post-image").firstOrNull()
            var cover = img?.attr("abs:data-src")?.ifEmpty { img.attr("abs:data-lazy-src")?.ifEmpty { img.attr("abs:src")?.ifEmpty { img.attr("src") } } } ?: ""
            if (cover.startsWith("data:image")) {
                cover = img?.attr("abs:data-src")?.ifEmpty { img.attr("abs:data-lazy-src") } ?: ""
            }
            if (cover.startsWith("//")) cover = "https:$cover"
            else if (cover.startsWith("/")) cover = "$baseUrl$cover"
            thumbnail_url = cover

            description = document.selectFirst("div[itemprop=description], div.entry-content, .synopsis, .description, .summary")?.text()?.trim()

            val genres = document.select("div.gnr a, a[href*='/genres/'], .genres-content a, .mgen a").map { it.text().trim() }.filter { it.isNotEmpty() }.distinct()
            genre = genres.joinToString(", ")

            author = document.select("div.imptdt:contains(Publicado por) i, div.imptdt:contains(Autor) i, div.spe span:contains(Autor)").firstOrNull()?.text()?.trim()
            artist = document.select("div.imptdt:contains(Artista) i, div.spe span:contains(Artista)").firstOrNull()?.text()?.trim()

            val statusText = document.select("div.imptdt:contains(Estado) i, div.spe span:contains(Estado)").firstOrNull()?.text()?.lowercase() ?: ""
            status = when {
                statusText.contains("ongoing") || statusText.contains("emisión") || statusText.contains("emision") || statusText.contains("curso") -> SManga.ONGOING
                statusText.contains("completed") || statusText.contains("complet") || statusText.contains("final") -> SManga.COMPLETED
                statusText.contains("hiatus") || statusText.contains("paus") -> SManga.ON_HIATUS
                statusText.contains("dropped") || statusText.contains("cancel") -> SManga.CANCELLED
                else -> SManga.ONGOING
            }
            initialized = true
        }
    }

    override fun chapterListParse(response: Response): List<SChapter> {
        val document = response.asJsoup()
        val chapters = mutableListOf<SChapter>()
        val seen = mutableSetOf<String>()

        val elements = document.select("ul.clstyle li, #chapterlist li, div.eplister li, li.wp-manga-chapter, div.mini-letters > a")
        for (el in elements) {
            val link = if (el.tagName() == "a") el else el.selectFirst("a") ?: continue
            var href = link.attr("abs:href").ifEmpty { link.attr("href") }
            if (href.isEmpty() || !seen.add(href)) continue
            if (href.startsWith("/")) href = "$baseUrl$href"

            val numAttr = el.attr("data-num")
            val nameSpan = el.selectFirst("span.chapternum, span.chapter-name")?.text()?.trim()
            val text = nameSpan ?: link.text().trim()

            val numMatch = Regex("""(\d+(\.\d+)?)""").find(if (numAttr.isNotEmpty()) numAttr else text)
                ?: Regex("""-capitulo-(\d+(\.\d+)?)""", RegexOption.IGNORE_CASE).find(href)
            val chNum = numMatch?.groupValues?.get(1)?.toFloatOrNull() ?: (chapters.size + 1).toFloat()

            val dateStr = el.selectFirst("span.chapterdate")?.text()?.trim()
            val dateUpload = try {
                if (!dateStr.isNullOrEmpty()) {
                    val relativeMatch = Regex("""(?:hace\s+)?(\d+)\s*(s|seg|min|m|h|hora|horas|d|dia|días|dias)""", RegexOption.IGNORE_CASE).find(dateStr)
                    if (relativeMatch != null) {
                        val amount = relativeMatch.groupValues[1].toLongOrNull() ?: 0L
                        val unit = relativeMatch.groupValues[2].lowercase()
                        val now = System.currentTimeMillis()
                        when {
                            unit.startsWith("s") -> now - (amount * 1000L)
                            unit.startsWith("m") -> now - (amount * 60 * 1000L)
                            unit.startsWith("h") -> now - (amount * 60 * 60 * 1000L)
                            unit.startsWith("d") -> now - (amount * 24 * 60 * 60 * 1000L)
                            else -> 0L
                        }
                    } else {
                        dateFormat.parse(dateStr)?.time ?: 0L
                    }
                } else 0L
            } catch (e: Exception) {
                0L
            }

            val chapter = SChapter.create().apply {
                setUrlWithoutDomain(href)
                this.name = if (text.isNotBlank() && text != href) text else "Capítulo ${if (chNum % 1f == 0f) chNum.toInt().toString() else chNum.toString()}"
                this.chapter_number = chNum
                this.date_upload = dateUpload
            }
            chapters.add(chapter)
        }

        return chapters.sortedByDescending { it.chapter_number }
    }

    override fun pageListParse(response: Response): List<Page> {
        val html = response.body.string()
        val pages = mutableListOf<Page>()
        val seen = mutableSetOf<String>()

        // 1. Parse JSON from ts_reader.run({...})
        val tsReaderMatch = Regex("""ts_reader\.run\s*\((.*?)\);""", RegexOption.DOT_MATCHES_ALL).find(html)
            ?: Regex("""\"images\"\s*:\s*\[([^\]]+)\]""").find(html)

        if (tsReaderMatch != null) {
            val imagesMatch = Regex("""\"images\"\s*:\s*\[([^\]]+)\]""").find(tsReaderMatch.value)
            if (imagesMatch != null) {
                val arrayContent = imagesMatch.groupValues[1]
                val urlMatches = Regex("""\"(https?:[^\"]+)\"""").findAll(arrayContent)
                for (m in urlMatches) {
                    var src = m.groupValues[1].replace("\\/", "/").trim()
                    if (src.isNotEmpty() && !src.contains("logo") && !src.contains("banner") && seen.add(src)) {
                        pages.add(Page(pages.size, "", src))
                    }
                }
            }
        }

        // 2. Fallback: Parse from HTML img elements
        if (pages.isEmpty()) {
            val document = org.jsoup.Jsoup.parse(html, baseUrl)
            val imgElements = document.select("div#readerarea img, div.reading-content img, .page-break img, .wp-manga-chapter-img")
            for (el in imgElements) {
                var src = el.attr("abs:data-src").ifEmpty { el.attr("abs:data-lazy-src").ifEmpty { el.attr("abs:src").ifEmpty { el.attr("src") } } }.trim()
                if (src.startsWith("data:image")) {
                    src = el.attr("abs:data-src").ifEmpty { el.attr("abs:data-lazy-src") }.trim()
                }
                if (src.startsWith("//")) src = "https:$src"
                else if (src.startsWith("/")) src = "$baseUrl$src"
                if (src.isNotEmpty() && !src.contains("logo") && !src.contains("banner") && !src.contains("readerarea.svg") && seen.add(src)) {
                    pages.add(Page(pages.size, "", src))
                }
            }
        }

        return pages
    }

    override fun imageRequest(page: Page): Request {
        val imgHeaders = headersBuilder()
            .set("Referer", "$baseUrl/")
            .set("Accept", "image/avif,image/webp,image/apng,image/svg+xml,image/*,*/*;q=0.8")
            .build()
        return GET(page.imageUrl!!, imgHeaders)
    }

    override fun imageUrlParse(response: Response): String = ""

    // --- Filters ---
    override fun getFilterList(): FilterList = FilterList(
        OrderByFilter(),
        StatusFilter(),
        TypeFilter(),
        GenreFilter(),
    )

    private class OrderByFilter : Filter.Select<String>(
        "Ordenar por",
        arrayOf("Más Vistos", "Últimos Actualizados", "Nuevos", "Alfabético (A-Z)"),
    ) {
        fun toUriPart(): String = when (state) {
            0 -> "popular"
            1 -> "update"
            2 -> "latest"
            3 -> "title"
            else -> "popular"
        }
    }

    private class StatusFilter : Filter.Select<String>(
        "Estado",
        arrayOf("Todos", "En emisión", "Completado", "Pausado", "Cancelado"),
    ) {
        fun toUriPart(): String = when (state) {
            1 -> "ongoing"
            2 -> "completed"
            3 -> "hiatus"
            4 -> "dropped"
            else -> ""
        }
    }

    private class TypeFilter : Filter.Select<String>(
        "Tipo",
        arrayOf("Todos", "Manhwa", "Manhua", "Manga", "Comic"),
    ) {
        fun toUriPart(): String = when (state) {
            1 -> "manhwa"
            2 -> "manhua"
            3 -> "manga"
            4 -> "comic"
            else -> ""
        }
    }

    private class GenreFilter : Filter.Select<String>(
        "Género",
        genres.map { it.first }.toTypedArray(),
    ) {
        fun toUriPart(): String = genres[state].second

        companion object {
            private val genres = arrayOf(
                Pair("Todos", ""),
                Pair("Acción", "accion"),
                Pair("Adulto", "adulto"),
                Pair("Artes Marciales", "artes-marciales"),
                Pair("Aventura", "aventura"),
                Pair("Ciencia Ficción", "ciencia-ficcion"),
                Pair("Comedia", "comedia"),
                Pair("Cultivación", "cultivacion"),
                Pair("Demonios", "demonios"),
                Pair("Deporte", "deporte"),
                Pair("Drama", "drama"),
                Pair("Ecchi", "ecchi"),
                Pair("Fantasía", "fantasia"),
                Pair("Girls Love", "girls-love"),
                Pair("Harem", "harem"),
                Pair("Histórico", "historico"),
                Pair("Horror", "horror"),
                Pair("Isekai", "isekai"),
                Pair("Josei", "josei"),
                Pair("Magia", "magia"),
                Pair("Mecha", "mecha"),
                Pair("Militar", "militar"),
                Pair("Misterio", "misterio"),
                Pair("Psicológico", "psicologico"),
                Pair("Realidad Virtual", "realidad-virtual"),
                Pair("Reencarnación", "reencarnacion"),
                Pair("Recuentos de la vida", "recuentos-de-la-vida"),
                Pair("Romance", "romance"),
                Pair("Seinen", "seinen"),
                Pair("Shounen", "shounen"),
                Pair("Shoujo", "shoujo"),
                Pair("Sistema", "sistema"),
                Pair("Sobrenatural", "sobrenatural"),
                Pair("Superpoderes", "superpoderes"),
                Pair("Tragedia", "tragedia"),
                Pair("Vampiros", "vampiros"),
                Pair("Venganza", "venganza"),
                Pair("Webtoon", "webtoon"),
                Pair("Yuri", "yuri"),
            )
        }
    }
}

